Connect sidebar to redux to receive site prop when available

allendav opened the pull request:
Fixes #14021 

To test, 

Starting at URL: calypso.localhost:3000/store/products/{site}/add
localStorage.clear();
indexedDB.deleteDatabase( 'calypso' );
refresh
store sidebar will fail to render initially
ensure sidebar does eventually render after a few seconds (when site id becomes available)

@kellychoffman - thoughts on what we should display while waiting for a site id?

cc @coderkevin

coderkevin commented:
@allendav The Calypso docs say we should try to display as much as possible, so maybe disabled menu items?

coderkevin commented on client/extensions/woocommerce/store-sidebar/index.js:
```diff
@@ -118,3 +123,17 @@ export default class StoreSidebar extends Component {
+	const selectedSiteId = getSelectedSiteId( state );
+	const isSingleSite = !! selectedSiteId || currentUser.site_count === 1;
+	const siteId = selectedSiteId || ( isSingleSite && getPrimarySiteId( state ) ) || null;
+	const site = getSite( state, siteId );
```
I wonder if we shouldn't think about putting this information at the base of our app, in order to keep the sidebar in sync with the content...

allendav replied:
Good idea - done (at least for the sidebar) in https://github.com/Automattic/wp-calypso/pull/14022/commits/2db9bb49f7dd460b832fee14dc3dadb9f7bd27fc

allendav commented:
@coderkevin wrote:

> @allendav The Calypso docs say we should try to display as much as possible, so maybe disabled menu items?

Agreed! Added disabled-ness in https://github.com/Automattic/wp-calypso/pull/14022/commits/2db9bb49f7dd460b832fee14dc3dadb9f7bd27fc

@jameskoster - please take a look at the styles I added for disabled-ness and let me know what you think - to see the disabled styles, follow the instructions at the top of this pull request

jameskoster commented:
Committed a change to use the placeholder animation seen in other areas of Calypso while 'waiting'.

![fading](https://cldup.com/gOOPOGCvQ1.gif)

The only downside to this is the awkward gap beneath the menu where the background is different. So perhaps we apply the `disabled` class to the main sidebar wrapper instead?

belcherj commented on client/extensions/woocommerce/index.js:
```diff
@@ -151,15 +155,23 @@ function addStorePage( storePage, storeNavigation ) {
+function mapStateToProps( state ) {
+	const currentUser = getCurrentUser( state );
+	const selectedSiteId = getSelectedSiteId( state );
+	const isSingleSite = !! selectedSiteId || currentUser.site_count === 1;
```
the logic of this and the next line is pretty complex.  Not sure how to make it better but took me a minute to reason through it.

allendav replied:
Yeah, its unfortunately standard fare to make sure we are looking at just one site

justinshreve replied:
Maybe worth creating a new selector or helper function so both this and [and the sidebar code](https://github.com/Automattic/wp-calypso/blob/cafb1a709743d0f4f589def09943373ec53513ba/client/my-sites/sidebar/sidebar.jsx#L599) could take use of it?

allendav replied:
Hey @justinshreve - good idea - I simplified it a bit in https://github.com/Automattic/wp-calypso/pull/14022/commits/05d496a23240468072a9f63f7b4fd431992c9203

jameskoster commented:
Been thinking about the disabled state some more and chatted with Kelly. 

Now considering the best approach to be using the placeholder animation, but use block placeholders as well, opposed to the actual text / icon. Kind of like what happens in the Reader;

![reader](https://cldup.com/6GeFUm5bme-1200x1200.png)

allendav commented:
@jameskoster - please take a look now - all placeholder-y goodness

justinshreve commented:
I dig the updated placeholder.

kellychoffman commented:
Loading animation looks slick! :shipit:

allendav commented:
@justinshreve @belcherj - good to go?

justinshreve commented on client/layout/sidebar/item.jsx:
```diff
@@ -37,7 +38,10 @@ export default React.createClass( {
 	render() {
 		const isExternalLink = isExternal( this.props.link );
-		const classes = classnames( this.props.className, { selected: this.props.selected } );
+		const classes = classnames( 'sidebar__item', this.props.className, {
```
Minor and optional but since this `render()` method uses quite a few props (and some more then once like `link`) it might be worth destructuring above so `this.props` doesn't need repeated so many times.

allendav replied:
Agreed! Fixed in https://github.com/Automattic/wp-calypso/pull/14022/commits/ef650dfd7aae2ba7ec9df10dc3bcc3f09e067143

justinshreve commented on client/layout/sidebar/item.jsx:
```diff
@@ -37,7 +38,10 @@ export default React.createClass( {
-		const classes = classnames( this.props.className, { selected: this.props.selected } );
+		const classes = classnames( 'sidebar__item', this.props.className, {
+			'is-placeholder': this.props.isPlaceholder,
+			selected: this.props.selected
```
Should include a trailing comma.

allendav replied:
Agreed! Fixed in https://github.com/Automattic/wp-calypso/pull/14022/commits/ef650dfd7aae2ba7ec9df10dc3bcc3f09e067143 - also switched to using property shorthand

justinshreve commented on client/state/current-user/selectors.js:
```diff
@@ -44,6 +44,21 @@ export function getCurrentUserLocale( state ) {
+		return null;
+	}
+
+	return user.site_count || null;
```
If we have a valid user but they actually don't have any sites (user only account), wouldn't `0` make more sense here?

allendav replied:
Agreed! Fixed in https://github.com/Automattic/wp-calypso/pull/14022/commits/ef650dfd7aae2ba7ec9df10dc3bcc3f09e067143

coderkevin replied:
Is that even possible? A site is created for a user by default on WPCOM.

justinshreve commented:
Tested again and works. I did notice that it will load the sidebar on a non-WooCommerce site with all the links (Add Product, etc). Maybe in a future PR we could just show some links that would lead people to the store setup flow so they could learn more?

allendav commented:
@justinshreve wrote:

> I did notice that it will load the sidebar on a non-WooCommerce site with all the links (Add Product, etc). Maybe in a future PR we could just show some links that would lead people to the store setup flow so they could learn more?

Yep - there is a separate issue for a "this site doesn't have woocommerce (yet)"

justinshreve commented:
🚢

coderkevin commented on client/extensions/woocommerce/index.js:
```diff
@@ -151,15 +155,23 @@ function addStorePage( storePage, storeNavigation ) {
+	if ( ! siteId && ( 1 === getCurrentUserSiteCount( state ) ) ) {
+		siteId = getPrimarySiteId( state );
+	}
+	const site = getSite( state, siteId );
```
If this is going to be the way the sidebar always gets the siteId, maybe we should standardize a selector on it so we get the same one. (If we don't do this, we'll end up with a populated sidebar and vacant pages)

allendav replied:
Agreed - I've added the selector in https://github.com/Automattic/wp-calypso/pull/14022/commits/3c0ec04fc96a5d46464a6d123899c97b94c4878a

coderkevin commented on client/extensions/woocommerce/index.js:
```diff
@@ -151,15 +155,23 @@ function addStorePage( storePage, storeNavigation ) {
 	renderWithReduxStore(
 		React.createElement(
-			StoreSidebar, {
+			connect( mapStateToProps )( StoreSidebar ), {
```
I don't know that I've ever seen HOC functions used inside of `createElement()` like this. Not sure how I feel about it.

allendav replied:
Um I did this because I thought you had a good suggestion here:

 https://github.com/Automattic/wp-calypso/pull/14022#discussion_r116342627

In order to pass the site into the sidebar as a prop, and have that prop come from state (and thereby keep the sidebar "in sync", is this not the only way?

coderkevin replied:
It was more the syntax of it. Having it all in one very long line of code. That's all.
When it's all strung together like this, it's easier to miss something while reading it. At least it's that way for me.

coderkevin replied:
On a closer look here, I think I'd recommend that the `connect()` statement be moved to `store-sidebar/index.js` as that's the way we normally use it. We just wrap it in `connect()` before we export.

allendav replied:
Fixed in https://github.com/Automattic/wp-calypso/pull/14022/commits/57501ef4662e51dec3431a8778c6a7c588560785

coderkevin commented on client/extensions/woocommerce/store-sidebar/store-ground-control.js:
```diff
@@ -27,7 +29,7 @@ const StoreGroundControl = ( { site, translate } ) => {
 				<Site
 					compact
-					site={ site }
+					site={ isPlaceholder ? null : site }
```
If `site === null` when it's a placeholder, you wouldn't have to have this ternary.

allendav replied:
Agreed. Fixed in https://github.com/Automattic/wp-calypso/pull/14022/commits/d31a67e864813f2aec6b237bd9de6f28c27c9a3f

allendav commented:
@belcherj @coderkevin - please approve and then i'll merge this bad boy

allendav commented:
@coderkevin - state is connected to props in the sidebar module itself again - let me know if this is good to go - https://github.com/Automattic/wp-calypso/pull/14022/commits/57501ef4662e51dec3431a8778c6a7c588560785

allendav commented:
Squashed

coderkevin commented on client/layout/sidebar/item.jsx:
```diff
@@ -37,31 +36,26 @@ export default React.createClass( {
 
 	render() {
-		const { className, children, icon, isPlaceholder, label, link,
-			onNavigate, postType, selected, tipTarget } = this.props;
```
Just curious why you removed this destructuring line in favor of repeating `this.props` several times below?

allendav replied:
I was putting the file back the way it exists on master
